### PR TITLE
Tenant health info upload

### PR DESCRIPTION
**Pull request by jinapurapu**
### What it does:
For registered clusters, after generating the Health Info report, Health Info is uploaded to Subnet, and latest metrics are visible in Subnet.

### What it looks like:
Screen Shot 2023-03-09 at 11 55 56 AM

Screen Shot 2023-02-03 at 4 26 25 PM
Screen Shot 2023-02-03 at 4 26 44 PM
Screen Shot 2023-01-26 at 2 04 02 PM
Screen Shot 2023-01-26 at 2 06 06 PM

### CODE REVIEW

File: restapi/admin_health_info.go [hunk 1]
```diff
@@ -108,12 +114,58 @@ func tarGZ(healthInfo interface{}, version string) ([]byte, error) {
...
 	if err != nil {
-		return nil, err
+		return nil, false, err
+	}
+	airgapMode := req.FormValue("airgap") == "true"
+	return &deadlineDuration, airgapMode, nil
+}
+
+func sendHealthInfoToSubnet(healthInfo interface{}, ctx context.Context, client MinioAdmin) error {
+	filename := fmt.Sprintf("%s-health_%s.json", "testing", time.Now().Unix())
```
Comment (kaankabalak) [inline restapi/admin_health_info.go:156]: Would it make sense to remove `testing` from the filename here?

File: portal-ui/src/screens/Console/HealthInfo/HealthInfo.tsx [hunk 2]
```diff
@@ -106,23 +108,34 @@ const HealthInfo = ({ classes }: IHealthInfo) => {
 
   const message = useSelector((state: AppState) => state.healthInfo.message);
 
-  const licenseInfo = useSelector(
-    (state: AppState) => state?.system?.licenseInfo
-  );
-
-  const { plan = "" } = licenseInfo || {};
-  const registeredCluster = plan === "STANDARD" || plan === "ENTERPRISE";
```
Comment (reivaj05) [inline portal-ui/src/screens/Console/HealthInfo/HealthInfo.tsx:114]: Why is this removed? is it a bug and licenseInfo is not stored in the slice?

Comment (jinapurapu) [reply portal-ui/src/screens/Console/HealthInfo/HealthInfo.tsx:114]: It stopped working from the slice suddenly, so I put in the workaround. I will try to debug getting licenseInfo from the slice again.

**Comment by dvaldivia** [comment]
Coverage is failing, you should add some tests for the new code @jinapurapu

File: restapi/admin_health_info.go [hunk 3]
```diff
@@ -51,14 +54,16 @@ func startHealthInfo(ctx context.Context, conn WSConn, client MinioAdmin, deadli
...
-
 	var err error
 	// Fetch info of all servers (cluster or single server)
 	healthInfo, version, err := client.serverHealthInfo(ctx, healthDataTypes, *deadline)
 	if err != nil {
 		return err
 	}
 
+	if !airgapMode {
+		sendHealthInfoToSubnet(ctx, healthInfo, client)
```
Comment (dvaldivia) [inline restapi/admin_health_info.go:65]: please handle this error

**Review by dvaldivia — CHANGES_REQUESTED** [review]

File: restapi/admin_health_info.go [hunk 4]
```diff
@@ -108,12 +113,58 @@ func tarGZ(healthInfo interface{}, version string) ([]byte, error) {
...
+		return nil, false, err
+	}
+	airgapMode := req.FormValue("airgap") == "true"
+	return &deadlineDuration, airgapMode, nil
+}
+
+func sendHealthInfoToSubnet(ctx context.Context, healthInfo interface{}, client MinioAdmin) error {
+	filename := fmt.Sprintf("health_%d.json", time.Now().Unix())
+
+	saveHealthInfo(healthInfo, filename)
```
Comment (dvaldivia) [inline restapi/admin_health_info.go:157]: please handle this error

File: restapi/admin_health_info.go [hunk 5]
```diff
@@ -108,12 +113,58 @@ func tarGZ(healthInfo interface{}, version string) ([]byte, error) {
...
+	subnetUploadURL := subnet.UploadURL("health", filename)
+	subnetHTTPClient := &xhttp.Client{Client: GetConsoleHTTPClient("")}
+	subnetTokenConfig, e := GetSubnetKeyFromMinIOConfig(ctx, client)
+	if e != nil {
+		return e
 	}
-	return &deadlineDuration, nil
+	apiKey := subnetTokenConfig.APIKey
+	headers := subnet.UploadAuthHeaders(apiKey)
+	subnet.UploadFileToSubnet(subnetHTTPClient, filename, subnetUploadURL, headers)
```
Comment (dvaldivia) [inline restapi/admin_health_info.go:166]: please handle this error

File: restapi/admin_health_info.go [hunk 6]
```diff
@@ -108,12 +113,58 @@ func tarGZ(healthInfo interface{}, version string) ([]byte, error) {
...
+	// Delete the file after successful upload
+	os.Remove(filename)
```
Comment (dvaldivia) [inline restapi/admin_health_info.go:168]: please handle this error

**Review by dvaldivia — CHANGES_REQUESTED** [review]

**Review by dvaldivia — CHANGES_REQUESTED** [review]
Please also show the generated URL page for the report

**Comment by prakashsvmx** [comment]
@jinapurapu  could you please resolve the conflicts ?

File: restapi/admin_health_info.go [hunk 7]
```diff
@@ -68,12 +71,19 @@ func startHealthInfo(ctx context.Context, conn WSConn, client MinioAdmin, deadli
...
 	}
 
+	subnetResp, err := sendHealthInfoToSubnet(ctx, healthInfo, client)
 	report := messageReport{
 		Encoded:          encodedDiag,
 		ServerHealthInfo: healthInfo,
+		SubnetResponse:   subnetResp,
+	}
+	if err != nil {
+		report.SubnetResponse = "ERROR"
```
Comment (reivaj05) [inline restapi/admin_health_info.go:84]: maybe fmt.Sprintf("Error: %s", err.Error()) ?

**Comment by allanrogerr** [comment]
@jinapurapu Also, when I hit "Sign Out", I face this blank screen. I effectively can't sign out.
image

**Comment by allanrogerr** [comment]
> @jinapurapu Thanks for your offline help! After registering my cluster I am facing this request to register my cluster still. How do I get past this? Let me know if you have any idea. I can't test unfortunately.

Resolved after pulling latest changes

**Comment by allanrogerr** [comment]
@jinapurapu New issue. As soon as I click "Start New Diagnostic", I face "An error occurred while getting the Diagnostic file."
image

**Review by allanrogerr — APPROVED** [review]
Works as expected

**Comment by allanrogerr** [comment]
> @jinapurapu Looks good. I ran diagnostics - observed the output - ran again and noticed that the value I changed did change. Except a bug unrelated to your change perhaps. I used to have 5 objects in my bucket. Then I deleted 2. However the UI is still reporting 5 objects (Access: PRIVATE102.6 MiB - 5 Objects)
> 
> See image
> 
> vs
> 
> image

Unrelated but valid issue

**Comment by jinapurapu** [comment]
> > Please also show the generated URL page for the report
> 
> @jinapurapu Was this addressed?

The link to the Subnet Health screen for the report is here - do we want to include the same display in Console Health as well?
Screen Shot 2023-03-16 at 12 55 16 PM

**Pull request merged**